### DISCUSSION TITLE
Application Logic

### DISCUSSION THREAD

**Discussion by afrankel** [Budibase/budibase#4259] [General]
Really nice piece of software. 

I am still a bit confused about a few items:
- application logic - I see automations but I can't connect the dots on how to do this. For example, let's say I have build a Budibase app that collects some data in a database table (rows) and I want to allow the user to click a button to download an xml version of that data and maybe validate it against an xsd file. Traditionally (in a server based model) I would create an API/function/method that would pull from the database and use a library such as xml-writer/xmllint/xml.js/... to do this. I can see there are automations and the use of Javascript, but how can I pull in a node library and what would be the steps here? How do I tie a button to this function and see the code? How can I return a file to the user?
- I see the html templates in the config, but I would have thought that I could create my own and then specify an html template when sending an email. I couldn't find this feature.
-  I assume there's no easy way to unit test the javascript code? 
- Can multiple developers work on a single piece of code? How would that work with a repository/git?

Many thanks

**Comment by mjashanks** [comment]
Thanks.

**Exporting data.**
Not possible at the moment, but we get asked for this a lot. It will get done, probably via a component in the design section.

**Using Javascript libraries**
We are hoping to add extensibility features in the first half of this year. I.e. building your own datasources and components. Not available right now.

**Custom Email Templates**
It's a good idea - we do not have this one either.

**Multiple developers**
By "A single piece of code", In assume that you mean "an app". Again, this is a feature that we hope to address in this half of the year. 

So - pretty much a no to everything right now, but you're on the same wavelength as us! These will all get done, it's just a matter of priorities.

Thanks for the feedback & suggestions.

**Reply by Cristiano-Ramos-Oliveira** [reply]
gostaria de solicita irformações de como efetuar castras de usuario de biblioteca

**Comment by afrankel** [comment]
Coming back to Budibase more than a year later I believe most/all have been implemented. Can you correct me if I'm wrong?:

- exporting: https://docs.budibase.com/docs/actions#export-data

- tying in some Javascript code when you press a button on a UI: https://docs.budibase.com/docs/actions#trigger-automation and https://docs.budibase.com/docs/js-scripting#adding-a-javascript-code-block

- adding javascript libraries (custom code) i.e., going outside of the box - either create a component: https://docs.budibase.com/docs/custom-component#getting-started or datasource:  https://docs.budibase.com/docs/custom-datasource#getting-started 

- custom email templates - as shown in step 5 here: https://docs.budibase.com/docs/js-scripting#tutorial-linking-an-api-response-to-table-rows

- multiple developers - ??

**Reply by ConorWebb96** [reply]
Hey @afrankel,

Yes everything you've listed has been implemented aside from the multiple developers.

By multiple developers, I'm assuming you are talking about having the ability to develop apps with more than 1 user at a time?

If thats the case, there is good news. This is currently being worked on and should be out within the next month or two!

I hope this answers your question!

**Reply by afrankel** [reply]
Excellent!
